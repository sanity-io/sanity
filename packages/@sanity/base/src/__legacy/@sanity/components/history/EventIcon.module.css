@import 'part:@sanity/base/theme/variables-style';

:root {
  --history-icon-size: calc(0.5em + 1px);
}

.root {
  position: relative;
  height: var(--history-icon-size);
  width: var(--history-icon-size);
  background: none;
  box-sizing: border-box;
  border-collapse: separate;
  opacity: 0.2;

  @nest [data-is-selected='true'] & {
    opacity: 1;
  }
}

.created {
  composes: root;
  border-radius: 50%;
  border: 1px solid currentColor;
}

.edited {
  composes: root;
  border-radius: 50%;
  border: 1px solid currentColor;
}

.published {
  composes: root;
  border-radius: 50%;
  background: currentColor;

  @nest [data-is-current-version='true'] & {
    opacity: 1;
    background-color: var(--state-success-color);
  }

  @nest [data-is-current-version='true'][data-is-selected='true'] & {
    border: 1px solid currentColor;
  }
}

.unpublished {
  composes: root;
  background: currentColor;

  @nest &::before {
    content: '';
    display: block;
    position: absolute;
    top: 0;
    left: calc(var(--history-icon-size) / 2 - 0.5px);
    bottom: 0;
    border-left: 1px solid currentColor;
  }
}

.truncated {
  composes: root;
}
