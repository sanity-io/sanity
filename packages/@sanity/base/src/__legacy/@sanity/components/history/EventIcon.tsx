import React from 'react'

import styles from './EventIcon.module.css'

interface EventIconProps {
  className: string
  type?: string
}

function EventIcon({className, type}: EventIconProps) {
  switch (type) {
    case 'created':
      return <div className={`${className} ${styles.created}`} aria-hidden="true" />
    case 'edited':
      return <div className={`${className} ${styles.edited}`} aria-hidden="true" />
    case 'published':
      return <div className={`${className} ${styles.published}`} aria-hidden="true" />
    case 'unpublished':
      return <div className={`${className} ${styles.unpublished}`} aria-hidden="true" />
    case 'truncated':
      return (
        <svg
          className={`${className} ${styles.truncated}`}
          aria-hidden="true"
          viewBox="0 0 9 9"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.97136 7.3637C7.14598 8.36308 5.89737 9 4.5 9C2.01472 9 0 6.98528 0 4.5C0 2.01472 2.01472 0 4.5 0C5.90844 0 7.16575 0.647052 7.99089 1.66011L3.5 4.59783L7.97136 7.3637Z"
            fill="currentColor"
          />
        </svg>
      )
    default:
      return <div className={`${className} ${styles.edited}`} aria-hidden="true" />
  }
}

export default EventIcon
