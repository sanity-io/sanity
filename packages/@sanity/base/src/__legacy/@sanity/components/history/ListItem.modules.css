@import 'part:@sanity/base/theme/variables-style';

:root {
  --history-circle-size: calc(0.5em + 1px);
  --history-line-top: var(--small-padding);
  --history-line-left: calc(var(--medium-padding) + 3px);
}

.root {
  display: block;
  text-decoration: none;
  position: relative;
  padding: var(--medium-padding);
  padding-top: calc(var(--small-padding) * 1.75);
  padding-left: calc(var(--medium-padding) * 2);
  box-sizing: border-box;
  width: 100%;
}

.unSelected {
  composes: root;
  composes: item from 'part:@sanity/base/theme/layout/selectable-style';
}

.selected {
  composes: root;
  composes: selected from 'part:@sanity/base/theme/layout/selectable-style';

  @nest & div[class^="AvatarCircle_root"] {
    box-shadow: 0 0 0 1px var(--selected-item-color);
  }

  @nest & div[class^="AvatarCircle_avatarInitials"] {
    color: var(--selectable-item-color--inverted);
  }
}

.disabled {
  composes: root;
}

.startLine {
  height: calc(var(--small-padding) + var(--history-line-top) - 1px);
  width: 0;
  position: absolute;
  top: 0;
  left: var(--history-line-left);
  border-left: 1px solid currentColor;
  opacity: 0.2;

  @nest .root:first-child & {
    display: none;
  }

  @nest .selected & {
    opacity: 1;
  }
}

.endLine {
  height: calc(
    100% - var(--small-padding) - var(--history-circle-size) - var(--history-line-top) + 1px
  );
  width: 0;
  position: absolute;
  top: calc(var(--small-padding) + var(--history-circle-size) + var(--history-line-top) - 1px);
  left: var(--history-line-left);
  border-left: 1px solid currentColor;
  opacity: 0.2;

  @nest .root:last-child & {
    display: none;
  }

  @nest .selected & {
    opacity: 1;
  }
}

.icon {
  position: absolute;
  top: calc(var(--small-padding) + var(--history-line-top) - 1px);
  left: calc(var(--history-line-left) - var(--history-circle-size) / 2 + 0.5px);
}

.status {
  text-transform: uppercase;
  font-size: var(--font-size-xsmall, 0.75rem);
  letter-spacing: 0.5px;
  line-height: 1;
}

.users {
  position: relative;
  display: inline-flex;
  align-items: center;
  font-size: var(--font-size-large);
}

.userIcons {
  display: flex;
}

.userAvatar {
  margin-right: -8px;
}

.userName {
  margin-left: var(--medium-padding);
  font-size: var(--font-size-xsmall, 0.75rem);
}

.title {
  font-weight: 600;
  margin-bottom: var(--small-padding);
  margin-top: 1px;
}

.children {
  padding: var(--medium-padding) 0;
}

.truncatedInfo {
  margin: var(--small-padding) 0 var(--medium-padding) 0;

  @nest & p {
    font-size: var(--font-size-xsmall, 0.75rem);
    margin: 0 0 var(--small-padding);
  }

  @nest .selected & a {
    color: inherit;
  }
}
