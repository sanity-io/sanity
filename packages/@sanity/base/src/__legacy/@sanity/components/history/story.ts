import {storiesOf} from 'part:@sanity/storybook'
import {withKnobs} from 'part:@sanity/storybook/addons/knobs'
import {ListStory} from './stories/list'
import {ListItemStory} from './stories/listItem'

storiesOf('@sanity/components/history', module)
  .addDecorator(withKnobs)
  .add('List', ListStory)
  .add('List item', ListItemStory)
