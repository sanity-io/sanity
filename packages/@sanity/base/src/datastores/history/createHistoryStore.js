import client from 'part:@sanity/base/client'
import {from, merge} from 'rxjs'
import {transactionsToEvents} from '@sanity/transaction-collator'
import {map, mergeMap, reduce, scan} from 'rxjs/operators'
import {getDraftId, getPublishedId} from '../../util/draftUtils'
import {omit, isUndefined} from 'lodash'
import jsonReduce from 'json-reduce'

const documentRevisionCache = Object.create(null)

const compileTransactions = (acc, curr) => {
  if (acc[curr.id]) {
    acc[curr.id].mutations = acc[curr.id].mutations.concat(curr.mutations)
    acc[curr.id].timestamp = curr.timestamp
  } else {
    acc[curr.id] = curr
  }
  return acc
}

const ndjsonToArray = (ndjson) => {
  return ndjson
    .toString('utf8')
    .split('\n')
    .filter(Boolean)
    .map((line) => JSON.parse(line))
}

const getHistory = (documentIds, options = {}) => {
  const ids = Array.isArray(documentIds) ? documentIds : [documentIds]
  const {time, revision} = options

  if (time && revision) {
    throw new Error(`getHistory can't handle both time and revision parameters`)
  }

  const dataset = client.clientConfig.dataset
  let url = `/data/history/${dataset}/documents/${ids.join(',')}`

  if (revision) {
    url = `${url}?revision=${revision}`
  } else {
    const timestamp = time || new Date().toISOString()
    url = `${url}?time=${timestamp}`
  }

  return client.request({url})
}

const getDocumentAtRevision = (documentId, revision) => {
  const publishedId = getPublishedId(documentId)
  const draftId = getDraftId(documentId)

  const cacheKey = `${publishedId}@${revision}`
  if (!(cacheKey in documentRevisionCache)) {
    const dataset = client.clientConfig.dataset
    const url = `/data/history/${dataset}/documents/${publishedId},${draftId}?revision=${revision}`
    documentRevisionCache[cacheKey] = client.request({url}).then(({documents}) => {
      const published = documents.find((res) => res._id === publishedId)
      const draft = documents.find((res) => res._id === draftId)
      return draft || published
    })
  }

  return documentRevisionCache[cacheKey]
}

const getTransactions = (documentIds) => {
  const ids = Array.isArray(documentIds) ? documentIds : [documentIds]
  const dataset = client.clientConfig.dataset
  const url = `/data/history/${dataset}/transactions/${ids.join(',')}?excludeContent=true`
  return client.request({url}).then(ndjsonToArray)
}

function historyEventsFor(documentId) {
  const pairs = [getDraftId(documentId), getPublishedId(documentId)]

  const query = '*[_id in $documentIds]'

  const pastTransactions$ = from(getTransactions(pairs)).pipe(
    mergeMap((transactions) => from(transactions)),
    map((trans) => ({
      author: trans.author,
      documentIDs: pairs,
      id: trans.id,
      mutations: trans.mutations,
      timestamp: trans.timestamp,
    })),
    reduce(compileTransactions, {})
  )

  const realtimeTransactions$ = client.observable.listen(query, {documentIds: pairs}).pipe(
    map((item) => ({
      author: item.identity,
      documentIDs: pairs,
      id: item.transactionId,
      mutations: item.mutations,
      timestamp: item.timestamp,
    })),
    scan(compileTransactions, {})
  )

  return merge(realtimeTransactions$, pastTransactions$).pipe(
    scan((prev, next) => {
      return {...prev, ...next}
    }, {}),
    map((transactions) =>
      transactionsToEvents(
        pairs,
        Object.keys(transactions).map((key) => transactions[key])
      ).reverse()
    )
  )
}

const getAllRefIds = (doc) =>
  jsonReduce(
    doc,
    (acc, node) =>
      node && typeof node === 'object' && '_ref' in node && !acc.includes(node._ref)
        ? [...acc, node._ref]
        : acc,
    []
  )

function jsonMap(value, mapFn) {
  if (Array.isArray(value)) {
    return mapFn(value.map((item) => jsonMap(item, mapFn)).filter((item) => !isUndefined(item)))
  }

  if (value && typeof value === 'object') {
    return mapFn(
      Object.keys(value).reduce((res, key) => {
        const mappedValue = jsonMap(value[key], mapFn)
        if (!isUndefined(mappedValue)) {
          res[key] = mappedValue
        }

        return res
      }, {})
    )
  }

  return mapFn(value)
}

const mapRefNodes = (doc, mapFn) =>
  jsonMap(doc, (node) => {
    return node && typeof node === 'object' && typeof node._ref === 'string' ? mapFn(node) : node
  })

export const removeMissingReferences = (doc, existingIds) =>
  mapRefNodes(doc, (refNode) => {
    const documentExists = existingIds[refNode._ref]
    return documentExists ? refNode : undefined
  })

function restore(id, targetId, rev) {
  return from(getDocumentAtRevision(id, rev)).pipe(
    mergeMap((documentAtRevision) => {
      const existingIdsQuery = getAllRefIds(documentAtRevision)
        .map((refId) => `"${refId}": defined(*[_id=="${refId}"]._id)`)
        .join(',')

      return client.observable
        .fetch(`{${existingIdsQuery}}`)
        .pipe(map((existingIds) => removeMissingReferences(documentAtRevision, existingIds)))
    }),
    map((documentAtRevision) =>
      // Remove _updatedAt and create a new draft from the document at given revision
      ({
        ...omit(documentAtRevision, '_updatedAt'),
        _id: targetId,
      })
    ),
    mergeMap((restoredDraft) =>
      client.observable.transaction().createOrReplace(restoredDraft).commit()
    )
  )
}

export default function createHistoryStore() {
  return {
    getDocumentAtRevision,
    getHistory,
    getTransactions,
    historyEventsFor,
    restore,
  }
}
