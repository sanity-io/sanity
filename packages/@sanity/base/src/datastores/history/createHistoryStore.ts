import {from, merge} from 'rxjs'
// eslint-disable-next-line import/no-unresolved
import {transactionsToEvents} from '@sanity/transaction-collator'
import {map, mergeMap, reduce, scan} from 'rxjs/operators'
import {omit, isUndefined} from 'lodash'
import jsonReduce from 'json-reduce'
import {getDraftId, getPublishedId} from '../../util/draftUtils'
import {versionedClient} from '../../client/versionedClient'

const documentRevisionCache = Object.create(null)

const compileTransactions = (acc, curr) => {
  if (acc[curr.id]) {
    acc[curr.id].mutations = acc[curr.id].mutations.concat(curr.mutations)
    acc[curr.id].timestamp = curr.timestamp
  } else {
    acc[curr.id] = curr
  }
  return acc
}

const ndjsonToArray = (ndjson) => {
  return ndjson
    .toString('utf8')
    .split('\n')
    .filter(Boolean)
    .map((line) => JSON.parse(line))
}

type HistoryTransaction = any

const getHistory = (documentIds, options: {time?: number; revision?: string} = {}) => {
  const ids = Array.isArray(documentIds) ? documentIds : [documentIds]
  const {time, revision} = options

  if (time && revision) {
    throw new Error(`getHistory can't handle both time and revision parameters`)
  }

  const dataset = versionedClient.clientConfig.dataset
  let url = `/data/history/${dataset}/documents/${ids.join(',')}`

  if (revision) {
    url = `${url}?revision=${revision}`
  } else {
    const timestamp = time || new Date().toISOString()
    url = `${url}?time=${timestamp}`
  }

  return versionedClient.request({url})
}

const getDocumentAtRevision = (documentId, revision) => {
  const publishedId = getPublishedId(documentId)
  const draftId = getDraftId(documentId)

  const cacheKey = `${publishedId}@${revision}`
  if (!(cacheKey in documentRevisionCache)) {
    const dataset = versionedClient.clientConfig.dataset
    const url = `/data/history/${dataset}/documents/${publishedId},${draftId}?revision=${revision}`
    documentRevisionCache[cacheKey] = versionedClient.request({url}).then(({documents}) => {
      const published = documents.find((res) => res._id === publishedId)
      const draft = documents.find((res) => res._id === draftId)
      return draft || published
    })
  }

  return documentRevisionCache[cacheKey]
}

const getTransactions = (documentIds) => {
  const ids = Array.isArray(documentIds) ? documentIds : [documentIds]
  const dataset = versionedClient.clientConfig.dataset
  const url = `/data/history/${dataset}/transactions/${ids.join(',')}?excludeContent=true`
  return versionedClient.request({url}).then(ndjsonToArray)
}

function historyEventsFor(documentId) {
  const pairs = [getDraftId(documentId), getPublishedId(documentId)]

  const query = '*[_id in $documentIds]'

  const pastTransactions$ = from(getTransactions(pairs)).pipe(
    mergeMap((transactions) => from(transactions)),
    map((trans: HistoryTransaction) => ({
      author: trans.author,
      documentIDs: pairs,
      id: trans.id,
      mutations: trans.mutations,
      timestamp: trans.timestamp,
    })),
    reduce(compileTransactions, {})
  )

  const realtimeTransactions$ = versionedClient.observable.listen(query, {documentIds: pairs}).pipe(
    map((item) => ({
      author: item.identity,
      documentIDs: pairs,
      id: item.transactionId,
      mutations: item.mutations,
      timestamp: item.timestamp,
    })),
    scan(compileTransactions, {})
  )

  return merge(realtimeTransactions$, pastTransactions$).pipe(
    scan((prev, next) => {
      return {...prev, ...next}
    }, {}),
    map((transactions) =>
      transactionsToEvents(
        pairs,
        Object.keys(transactions).map((key) => transactions[key])
      ).reverse()
    )
  )
}

const getAllRefIds = (doc) =>
  jsonReduce(
    doc,
    (acc, node) =>
      node && typeof node === 'object' && '_ref' in node && !acc.includes(node._ref)
        ? [...acc, node._ref]
        : acc,
    []
  )

function jsonMap(value, mapFn) {
  if (Array.isArray(value)) {
    return mapFn(value.map((item) => jsonMap(item, mapFn)).filter((item) => !isUndefined(item)))
  }

  if (value && typeof value === 'object') {
    return mapFn(
      Object.keys(value).reduce((res, key) => {
        const mappedValue = jsonMap(value[key], mapFn)
        if (!isUndefined(mappedValue)) {
          res[key] = mappedValue
        }

        return res
      }, {})
    )
  }

  return mapFn(value)
}

const mapRefNodes = (doc, mapFn) =>
  jsonMap(doc, (node) => {
    return node && typeof node === 'object' && typeof node._ref === 'string' ? mapFn(node) : node
  })

export const removeMissingReferences = (doc, existingIds) =>
  mapRefNodes(doc, (refNode) => {
    const documentExists = existingIds[refNode._ref]
    return documentExists ? refNode : undefined
  })

function restore(id, targetId, rev) {
  return from(getDocumentAtRevision(id, rev)).pipe(
    mergeMap((documentAtRevision) => {
      const existingIdsQuery = getAllRefIds(documentAtRevision)
        .map((refId) => `"${refId}": defined(*[_id=="${refId}"]._id)`)
        .join(',')

      return versionedClient.observable
        .fetch(`{${existingIdsQuery}}`)
        .pipe(map((existingIds) => removeMissingReferences(documentAtRevision, existingIds)))
    }),
    map((documentAtRevision) =>
      // Remove _updatedAt and create a new draft from the document at given revision
      ({
        ...omit(documentAtRevision, '_updatedAt'),
        _id: targetId,
      })
    ),
    mergeMap((restoredDraft: any) =>
      versionedClient.observable.transaction().createOrReplace(restoredDraft).commit()
    )
  )
}

export default function createHistoryStore() {
  return {
    getDocumentAtRevision,
    getHistory,
    getTransactions,
    historyEventsFor,
    restore,
  }
}
