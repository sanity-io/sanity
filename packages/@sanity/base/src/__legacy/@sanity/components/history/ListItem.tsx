import React from 'react'
import EventIcon from './EventIcon'

import styles from './ListItem.modules.css'

interface HistoryListItemProps {
  status?: string
  title?: string
  children?: React.ReactNode
  isCurrentVersion?: boolean
  isSelected?: boolean
  onSelect: (evt: React.MouseEvent<HTMLDivElement>) => void
  onEnterKey: () => void
  onArrowUpKey: () => void
  onArrowDownKey: () => void
  rev?: string
  tooltip?: string
  type?: string
  users?: {
    name?: string
    email?: string
    imageUrl?: string
    id?: string
  }[]
  linkParams?: Record<string, unknown>
  linkComponent?: React.ComponentType<{params: Record<string, unknown>}>
}

const noop = () => undefined

export default class HistoryListItem extends React.PureComponent<HistoryListItemProps> {
  static defaultProps = {
    status: 'Edited',
    title: undefined,
    onSelect: noop,
    onEnterKey: noop,
    onArrowUpKey: noop,
    onArrowDownKey: noop,
    isCurrentVersion: false,
    isSelected: false,
    users: [],
    children: undefined,
    rev: undefined,
    linkParams: undefined,
    linkComponent: undefined,
  }

  _rootElement: React.RefObject<HTMLDivElement> = React.createRef()

  componentDidUpdate(prevProps) {
    const {isSelected} = this.props

    // Focus the element when it becomes selected
    if (isSelected && !prevProps.isSelected) {
      this.focus()
    }
  }

  focus() {
    if (this._rootElement && this._rootElement.current) {
      this._rootElement.current.focus()
    }
  }

  handleKeyUp = (event: React.KeyboardEvent<HTMLDivElement>) => {
    const {onEnterKey} = this.props
    if (event.key === 'Enter') {
      onEnterKey()
    }
  }

  handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    // Prevent arrow keypress scrolling
    const {onArrowUpKey, onArrowDownKey} = this.props
    if (event.key === 'ArrowDown') {
      onArrowDownKey()
      event.preventDefault()
    } else if (event.key === 'ArrowUp') {
      onArrowUpKey()
      event.preventDefault()
    }
  }

  handleSelect = (evt: React.MouseEvent<HTMLDivElement>) => {
    this.props.onSelect(evt)
  }

  // eslint-disable-next-line complexity
  render() {
    const {
      linkComponent,
      linkParams,
      status,
      isSelected,
      title,
      children,
      isCurrentVersion,
      rev,
      tooltip,
      type,
    } = this.props
    const selectionClassName = isSelected ? styles.selected : styles.unSelected

    const content = (
      <>
        <EventIcon className={styles.icon} type={type} />
        <div className={styles.startLine} aria-hidden="true" />
        <div className={styles.endLine} aria-hidden="true" />
        <div className={styles.status}>{status}</div>
        {title && type !== 'truncated' && <div className={styles.title}>{title}</div>}
        {type === 'truncated' && (
          <div className={styles.truncatedInfo}>
            <p>
              <a
                href="https://www.sanity.io/docs/content-studio/history-experience"
                target="_blank"
                rel="noopener noreferrer"
              >
                Learn about history retention
              </a>
            </p>
          </div>
        )}

        {children && <div className={styles.children}>{children}</div>}
      </>
    )

    const rootProps = {
      className: selectionClassName,
      'data-type': type,
      'data-is-current-version': isCurrentVersion,
      'data-is-selected': isSelected,
      'data-rev': rev,
      tabIndex: type === 'truncated' ? undefined : 0,
      onKeyUp: this.handleKeyUp,
      onKeyDown: this.handleKeyDown,
      title: tooltip,
      ref: this._rootElement,
    }

    const ParameterizedLink = linkComponent
    const linkRev = isCurrentVersion ? '-' : rev

    return ParameterizedLink ? (
      <ParameterizedLink params={{...linkParams, rev: linkRev}} {...rootProps}>
        {content}
      </ParameterizedLink>
    ) : (
      <div {...rootProps} onClick={this.handleSelect}>
        {content}
      </div>
    )
  }
}
