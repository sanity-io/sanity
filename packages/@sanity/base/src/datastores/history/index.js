import createHistoryStore from './createHistoryStore'

export default createHistoryStore()
