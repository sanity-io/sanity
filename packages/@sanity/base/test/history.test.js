import {removeMissingReferences} from '../src/datastores/history/createHistoryStore'

const testDoc = {
  _id: 'foo',
  _type: 'test',
  string: 'value',
  number: 123,
  bool: true,
  subObject: {
    _type: 'sub',
    nested: {
      prop: true,
      array: [
        'contents',
        {_type: 'reference', _ref: 'abc123'},
        {_type: 'nonref', someProp: 'yes-it-exists'}
      ]
    }
  },
  arrayOfStrings: ['a', 'b', 'c'],
  arrayOfObjects: [
    {_type: 'foo', prop: 'yes'},
    {_type: 'reference', _ref: 'random'},
    {_type: 'foo', prop: 'no'},
    {_type: 'reference', _ref: 'd987abc'}
  ]
}

describe('removeMissingReferences', () => {
  test('removes references to missing docs, deeply', () => {
    const existingIds = {abc123: true, d987abc: false}
    const mapped = removeMissingReferences(testDoc, existingIds)
    expect(mapped).toMatchSnapshot()
  })
})
