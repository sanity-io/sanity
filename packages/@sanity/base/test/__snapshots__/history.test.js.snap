// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeMissingReferences removes references to missing docs, deeply 1`] = `
Object {
  "_id": "foo",
  "_type": "test",
  "arrayOfObjects": Array [
    Object {
      "_type": "foo",
      "prop": "yes",
    },
    Object {
      "_type": "foo",
      "prop": "no",
    },
  ],
  "arrayOfStrings": Array [
    "a",
    "b",
    "c",
  ],
  "bool": true,
  "number": 123,
  "string": "value",
  "subObject": Object {
    "_type": "sub",
    "nested": Object {
      "array": Array [
        "contents",
        Object {
          "_ref": "abc123",
          "_type": "reference",
        },
        Object {
          "_type": "nonref",
          "someProp": "yes-it-exists",
        },
      ],
      "prop": true,
    },
  },
}
`;
